### PR TITLE
Introduce environment variables

### LINKED ISSUE #23242: Add support for configuration through environment variables
#### Is your feature request related to a problem?

Configuration information like database credentials currently needs to be stored in a file within the PrestaShop install. This makes deployment more difficult and could be a security threat.

#### Describe the solution you'd like

Add support for using environment variables for all configurations (database credentials, etc)
The first step for this issue is to make all configurations variables generated in `parameters.php` usable with environment variables:

- database_host
- database_port
- database_name
- database_user
- database_password
- database_prefix
- database_engine
- mailer_transport
- mailer_host
- mailer_user
- mailer_password
- secret
- ps_caching
- ps_cache_enable
- ps_creation_date
- locale
- use_debug_toolbar
- cookie_key
- cookie_iv
- new_cookie_key

Be aware that there are parameter set on the legacy code ⚠️ 

(@Quetzacoalt91 please complete as needed).

### PR DESCRIPTION
**Pull request by MeKeyCool**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Define env variables and use them if they are set
| Type?             | new feature
| Category?         | IN
| BC breaks?        | yes
| Deprecations?     | no
| Fixed ticket?     | Fixes #23242.
| How to test?      | Define env variables in .env file, on install, check "Use env variables if set" and see that the env variables are used.
| Possible impacts? | -

## BC Break

```Install::processParameters(array $parameters, bool $useEnvValuesIfExists)``` now takes a second parameter ```$useEnvValuesIfExists``` as ```bool```

> :point_up: This Pull Request replaces https://github.com/PrestaShop/PrestaShop/pull/26077

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27990)

### CODE REVIEW

File: src/Core/Util/ConfigurationVariablesLoader.php
Comment (matks) [inline src/Core/Util/ConfigurationVariablesLoader.php]:
```suggestion
    public function loadEnvironmentVariables(array $parameters): array
```

What do you think? Do we use full name or short name?

**Review by matks — DISMISSED** [review]
OK for me 👍 I added a nice to have comment

When this PR is merged we'll need to document this feature 😉 on the [docs](https://github.com/prestashop/docs) repository

File: src/Core/Util/ConfigurationVariablesLoader.php
Comment (MeKeyCool) [reply src/Core/Util/ConfigurationVariablesLoader.php]: I'll change this. With auto completion in most of IDE, sounds better to have full names.

Comment (matks) [reply src/Core/Util/ConfigurationVariablesLoader.php]: Agreed 👍 

Also short names like "env", "conf", "Mgr" can be confusing especially for junior developers

**Review by matks — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

File: app/config/set_parameters.php
Comment (PululuK) [inline app/config/set_parameters.php]:
```suggestion
if (isset($container) && $container instanceof ContainerInterface) {
```

File: app/config/set_parameters.php
Comment (PululuK) [inline app/config/set_parameters.php]:
```suggestion
use Symfony\Component\DependencyInjection\ContainerInterface;
```

Comment (matks) [reply app/config/set_parameters.php]: Nice suggestion 👍

**Review by matks — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]
@MeKeyCool Thank you for your contribution. I don't see how in CLI install, the .env file is used.

**Comment by MeKeyCool** [comment]
> @MeKeyCool Thank you for your contribution. I don't see how in CLI install, the .env file is used.

As discussed with @eternoendless and @matks , this is not that easy and we could split this task and the CLI configuration.

I tried to create a feature request #28038 . If you agree, we may continue discussing and building specifications there :)

**Review by Progi1984 — DISMISSED** [review]
I disagree.

The scope is defined there : 
> Configuration information like database credentials currently needs to be stored in a file within the PrestaShop install.

The install is done in CLI / HTTP.

**Comment by PierreRambaud** [comment]
> I disagree.
> 
> The scope is defined there :
> 
> > Configuration information like database credentials currently needs to be stored in a file within the PrestaShop install.
> 
> The install is done in CLI / HTTP.

More than this, when you deploy an application, you better do install it with the CLI instead of UI.

**Comment by matks** [comment]
@Progi1984 @PierreRambaud The scope of https://github.com/PrestaShop/PrestaShop/issues/23242 has been reduced to HTTP only in order to allow delivering small scoped PRs.

The use of environment variables in CLI is relevant which is why https://github.com/PrestaShop/PrestaShop/issues/28038 was created. A PR to enable using environment variables in CLI will be delivered to fix https://github.com/PrestaShop/PrestaShop/issues/28038.

**Comment by Progi1984** [comment]
I see nothing in #23242 where it has been decided. But if you have decided, I remove my "Request Changes".

**Review by atomiix — DISMISSED** [review]

**Comment by matks** [comment]
@MeKeyCool I think this PR should be tested by a developer, wdyt?

Could you provide a list of steps as clear as possible to test this?

**Comment by MeKeyCool** [comment]
> @MeKeyCool I think this PR should be tested by a developer, wdyt?
> 
> Could you provide a list of steps as clear as possible to test this?

You are perfectly right.
Better than step to reproduce, the tester may test the documentation as well.
I expect to finish this documentation in coming days and come back to this task to explain expected behavior.

**Review by MeKeyCool — COMMENTED** [review]
Hi there ! I am currently working on documentation for this feature but I have some questions.

- `PS_CREATION_DATE` => I doubt about the interest to make this variable editable from environment ? What do you think @eternoendless ?
- @PrestaShop/qa-team, are you able / do you have enough rights to test editing files on PS server ? And if you can, do you have several database servers or emailers to test configuration ? Do you have any tool or access to validate cache behavior ?
- I was looking for existing documentation to avoid duplication. About configuration, i found that : https://devdocs.prestashop.com/1.7/basics/installation/configuration . And I know that there is a cli command for install : https://github.com/PrestaShop/PrestaShop/blob/develop/.github/workflows/php.yml#L138 and I found some information about configuration there : https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/classes/datas.php but nothing in devdoc. I found this install guide : https://doc.prestashop.com/display/PS17/Installing+PrestaShop+using+the+command-line+script for detailed options. I think @eternoendless you told me something about this documentation duplication but it is not clear for me (I suppose that doc.prestashop.com must disappear but I am not sure). Does anybody knows if something else exists ?

**Comment by matks** [comment]
> I was looking for existing documentation to avoid duplication. About configuration, i found that : https://devdocs.prestashop.com/1.7/basics/installation/configuration . And I know that there is a cli command for install : https://github.com/PrestaShop/PrestaShop/blob/develop/.github/workflows/php.yml#L138 and I found some information about configuration there : https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/classes/datas.php but nothing in devdoc. I found this install guide : https://doc.prestashop.com/display/PS17/Installing+PrestaShop+using+the+command-line+script for detailed options. I think @eternoendless you told me something about this documentation duplication but it is not clear for me (I suppose that doc.prestashop.com must disappear but I am not sure). Does anybody knows if something else exists ?

https://devdocs.prestashop.com/ is the Developer documentation it's in English only
https://doc.prestashop.com/ is the User documentation it is supposed to be read by not-technical users, it's translated in multiple languages

It seems like there is no "How to install PrestaShop using CLI" page in devdocs. For me it's an issue.

I suggest
- to create a new devdocs page in https://devdocs.prestashop.com/1.7/basics/installation/
- to copy (yes, duplicate 😛 ) the [content](https://doc.prestashop.com/display/PS17/Installing+PrestaShop+using+the+command-line+script) you found about CLI install because it's relevant inside this page
- then to add information about environment variables at the bottom of the page

**Comment by MeKeyCool** [comment]
@eternoendless / @PrestaShop/product-team , I created an issue about what I would like to do on documentation : https://github.com/PrestaShop/docs/issues/1365
I am waiting for comments / approval on it ^^

Ps. please don't miss my question about `PS_CREATION_DATE` ;)

**Comment by MeKeyCool** [comment]
@matks , It looks this PR is stuck.

I would like to remove `PS_CREATION_DATE` from the scope and I'll open a PR with https://github.com/PrestaShop/docs/issues/1365 suggestions applied to devdoc.

What do you think ? 
Should I wait more ? Is it ok to invest time to push some change ? 
Do you think I don't have more answer / comments cause people are too busy ? They don't care / feel concerned by this subject ? They don't understand / I am not clear and I should reword my questions / suggestions ? ...

**Comment by matks** [comment]
@MeKeyCool You're right, this PR must not be stuck because people cannot take time to answer.

So I'll go ahead and provide a clear set of steps so we can move forward :

1. **This PR is approved for code review. We just need to verify it behaves correctly.**
Can you provide a small paragraph explaining what's supposed to happen, what is the expected behavior and **I'll test it**.

2. **Let's remove PS_CREATION_DATE from the scope.** I take the decision and the responsibility if later is it found it was a mistake.

Is that OK?

**Comment by MeKeyCool** [comment]
> @MeKeyCool You're right, this PR must not be stuck because people cannot take time to answer.
> 
> So I'll go ahead and provide a clear set of steps so we can move forward :
> 
>     1. **This PR is approved for code review. We just need to verify it behaves correctly.**
>        Can you provide a small paragraph explaining what's supposed to happen, what is the expected behavior and **I'll test it**.
> 
>     2. **Let's remove PS_CREATION_DATE from the scope.** I take the decision and the responsibility if later is it found it was a mistake.
> 
> 
> Is that OK?

Reading deeper the source code, I think the implemented behavior is

| Variable set in `parameters.yaml` | `.env` or `.env.{ps_env}` file exists | Variable exists in `.env`/`.env.{ps_env}` but not set | Variable defined in `.env`/`.env.{ps_env}` | Variable defined in environment | Prestashop used variable    |
| --------------------------------- | ------------------------------------- | ----------------------------------------------------- | ------------------------------------------ | ------------------------------- | --------------------------- |
| Y                                 | Y                                     | Y                                                     | Y                                          | Y                               | from `.env`/`.env.{ps_env}` |
| Y                                 | Y                                     | Y                                                     | Y                                          | N                               | from `.env`/`.env.{ps_env}` |
| Y                                 | Y                                     | Y                                                     | N                                          | Y                               | from environment            |
| Y                                 | Y                                     | Y                                                     | N                                          | N                               | from parameters.yml         |
| Y                                 | Y                                     | N                                                     | N                                          | Y                               | from parameters.yml         |
| Y                                 | Y                                     | N                                                     | N                                          | N                               | from parameters.yml         |
| Y                                 | N                                     | N                                                     | N                                          | Y                               | from parameters.yml         |
| Y                                 | N                                     | N                                                     | N                                          | N                               | from parameters.yml         |

> :point_up:  We can find some information there :  https://symfony.com/doc/4.3/components/dotenv.html

Is it alright for you ?
I don't like the need to have variables defined in an empty env file to use environment variables but it is not blocking for me if it is expected behavior.

About testing, after few discussions with @sowbiba and @matks , it seems that the way to test would be to change `PS_DATABASE_NAME` and see what database prestashop uses.

I would like to ensure a better DoD and a devops friendlier way to get prestashop state.

File: tests/Unit/Core/Util/ConfigurationVariablesLoaderTest.php
Comment (atomiix) [inline tests/Unit/Core/Util/ConfigurationVariablesLoaderTest.php:177]: I guess this file needs to be removed at the end of the test?

Comment (MeKeyCool) [reply tests/Unit/Core/Util/ConfigurationVariablesLoaderTest.php:177]: Yep, this test should be reviewed to manage filesystem properly I guess :thinking:

File: src/Core/Util/ConfigurationVariablesLoader.php
Comment (Progi1984) [inline src/Core/Util/ConfigurationVariablesLoader.php:35]: Why define it as private as the class is overridable ?

File: tests/Unit/Core/Util/ConfigurationVariablesLoaderTest.php
Comment (Progi1984) [inline tests/Unit/Core/Util/ConfigurationVariablesLoaderTest.php:149]:
```suggestion
    private function createBasicEnvFile(): string
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Check comments, please :)

**Comment by MeKeyCool** [comment]
> Check comments, please :)

Thanks for reviewing @Progi1984 but with @matks we discussed yesterday and we are thinking to delay this task/issue cause it looks it is not that easy and expected behavior / scope isn't clear enough ^^' ...

(I should add some comments to https://github.com/PrestaShop/PrestaShop/issues/23242 today)

**Comment by Progi1984** [comment]
@MeKeyCool Sorry but I didn't see your conversation anywhere. Please don't discuss it in private, we are in open source project. Personnally, the extendability is not validated with `private const`. So I keep my Request Changes.

**Comment by MeKeyCool** [comment]
> @MeKeyCool Sorry but I didn't see your conversation anywhere. Please don't discuss it in private, we are in open source project. Personnally, the extendability is not validated with `private const`. So I keep my Request Changes.

You are perfectly right and I feel sorry for that situation (my comment was just a way to introduce the subject evolution ^^').
The discussion occurred late yesterday and I'll share everything today ^^'
Thanks for your comments and your time.

**Comment by MeKeyCool** [comment]
I added some comments to #23242 and I'll probably close this PR soon.

**Pull request closed without merging**
